### DISCUSSION TITLE
JSBSim on RISC-V

### DISCUSSION THREAD

**Discussion by dpculp** [JSBSim-Team/jsbsim#835] [General]
I've got JSBSim running on a RISC-V based single board computer.  The SBC is the new VisionFive 2 from StarFive.  It has a quad core CPU, 8 GB of memory, and runs Debian linux.  The computer is not very fast.  Even the Xfce desktop runs slowly, but considering how new the tech is I'm not disappointed. 

The Debian image supplied by StarFive contains the gcc version 11.3.0 compiler and Python 3.10.5

I've had no trouble getting other software built for the RISC-V 64 bit platform using apt.  Things like cmake, git, gnome system tools, htop, gedit.

![VisionFive2](https://user-images.githubusercontent.com/21369577/218240370-29df58a5-0415-46c2-b8bc-dd88fa03e976.jpg)

![VF2-screen-shot](https://user-images.githubusercontent.com/21369577/218240539-5d9f0e59-833c-4685-a94a-7a11437dd03a.jpg)

**Comment by bcoconni** [comment]
Nice !

**Comment by dpculp** [comment]
I suspected JSBSim would run OK considering there is already a normal linux system running on the board.  What I'm still not sure of is how the hardware timer compares to the timer on ARM or x86 chips.  I don't have test plan for that, but I'll look into it.  In the mean time I'll source a heat sink and fan before doing any stress testing.

**Comment by seanmcleod** [comment]
@dpculp out of interest, did you buy this SBC specifically for JSBSim use, or are you planning on using it for something else and just decided to check out JSBSim on it as well?

**Reply by dpculp** [reply]
@seanmcleod I originally ordered it out of general interest.  AFAIK it's the first RISC-V SBC that can be used as a desktop computer.  While it was on the way I wondered about JSBSim and how the hardware timer would work compared to other archs.  Now that I have the board  set up and have built some apps with it I realize I don't have a good way to test the timer.

Next test will be to build QtJSBSim for it.  I've read on the forums that Qt has been ported to the RISC-V-64 except for the QWebEngine module.  This will also be a test of SDL2 and joystick support.

**Comment by seanmcleod** [comment]
> I wondered about JSBSim and how the hardware timer would work compared to other archs.

Are you referring to the timer that JSBSim makes use of from the OS for the realtime path?

https://github.com/JSBSim-Team/jsbsim/blob/08a3ae58fe9ead2bdab786692aec2119ad8f3d63/src/JSBSim.cpp#L530-L555

In other words `getcurrentseconds()` and possibly also `sim_nsleep()`?

https://github.com/JSBSim-Team/jsbsim/blob/08a3ae58fe9ead2bdab786692aec2119ad8f3d63/src/JSBSim.cpp#L112-L144

**Reply by bcoconni** [reply]
A bit off topic but the `__BORLANDC__` flag is quite outdated :smile:
May be we should get rid of it, one day ?

**Reply by dpculp** [reply]
Yes,  I remember looking into timers a few years ago while writing QtJSBSim, and I saw some hardware variations, but it seems the kernel hides the differences from us by rounding.  According to this:  [https://linux.die.net/man/7/time](https://linux.die.net/man/7/time) the newer kernels, 2.5 and 2.6 have finer resolution available.

I think I'd have to run in real time at a high refresh rate to see any difference between architectures, and even then with the new kernels there might be no difference.  It's an experiment for a later time.

From the time(7) man page linked above:

`The hardware clock

Most computers have a (battery-powered) hardware clock which the kernel reads at boot time in order to initialize the software clock. For further details, see [rtc](https://linux.die.net/man/4/rtc)(4) and [hwclock](https://linux.die.net/man/8/hwclock)(8).
The software clock, HZ, and jiffies

The accuracy of various system calls that set timeouts, (e.g., [select](https://linux.die.net/man/2/select)(2), [sigtimedwait](https://linux.die.net/man/2/sigtimedwait)(2)) and measure CPU time (e.g., [getrusage](https://linux.die.net/man/2/getrusage)(2)) is limited by the resolution of the software clock, a clock maintained by the kernel which measures time in jiffies. The size of a jiffy is determined by the value of the kernel constant HZ.
The value of HZ varies across kernel versions and hardware platforms. On i386 the situation is as follows: on kernels up to and including 2.4.x, HZ was 100, giving a jiffy value of 0.01 seconds; starting with 2.6.0, HZ was raised to 1000, giving a jiffy of 0.001 seconds. Since kernel 2.6.13, the HZ value is a kernel configuration parameter and can be 100, 250 (the default) or 1000, yielding a jiffies value of, respectively, 0.01, 0.004, or 0.001 seconds. Since kernel 2.6.20, a further frequency is available: 300, a number that divides evenly for the common video frame rates (PAL, 25 HZ; NTSC, 30 HZ).

The [times](https://linux.die.net/man/2/times)(2) system call is a special case. It reports times with a granularity defined by the kernel constant USER_HZ. User-space applications can determine the value of this constant using sysconf(_SC_CLK_TCK).

High-resolution timers

Before Linux 2.6.21, the accuracy of timer and sleep system calls (see below) was also limited by the size of the jiffy.
Since Linux 2.6.21, Linux supports high-resolution timers (HRTs), optionally configurable via CONFIG_HIGH_RES_TIMERS. On a system that supports HRTs, the accuracy of sleep and timer system calls is no longer constrained by the jiffy, but instead can be as accurate as the hardware allows (microsecond accuracy is typical of modern hardware). You can determine whether high-resolution timers are supported by checking the resolution returned by a call to [clock_getres](https://linux.die.net/man/2/clock_getres)(2) or looking at the "resolution" entries in /proc/timer_list.

HRTs are not supported on all hardware architectures. (Support is provided on x86, arm, and powerpc, among others.)`

**Comment by jonsberndt** [comment]
Bertrand wrote:

A bit off topic but the __BORLANDC__ flag is quite outdated 😄
May be we should get rid of it, one day ?

I was noticing this recently – I agree, this doesn’t really apply any more and is a minor nuisance. The flag can almost surely be removed and the code cleaned up.

JB

**Comment by dpculp** [comment]
I ran a tougher test using a 500 hz integration rate, thinking the higher rate (which is also not divisible by 30) might induce a timer round-off.  On both of the machines, an amd64 and a riscv64, the results after 5 minutes of a real time run are the same.

My fears of a hardware timer difference were unfounded.  Or, the test isn't testing what I think it's testing.
